ls1intum/Artemis
# [Migration] Metis Data Model: Introduce Post Tags, Course-Wide Context and Reactions

### Checklist

### Motivation and Context


This is the 2. PR towards integration a new data model for Metis. With the introduction of new fields in the classes `Post` and `AnswerPost` prepare for implementing the new features in the following.

### Description

Three main parts of the PR are:
1. Migration: add new fields (`tags`, `reactions`) and tables (`Reaction`)
2. Refactor Server test: include new fields, improve test quality, reduce code duplication
3. Introduce endpoints for reaction: CREATE, DELETE
4. Add custom validations for `PostContextContraintValidator` and `ReactionConstraintValidator`
5. Added missinng client types `posting.model.ts` (common attributes of post and answer post), `reaction.model.ts`

The underlying data model is visualized here (in case this helps for reviewing or you are interested)
<img width="909" alt="Bildschirmfoto 2021-06-14 um 08 00 24" src="https://user-images.githubusercontent.com/41366522/121846016-9cca9a80-cce6-11eb-9e8a-814e66abd4d9.png">

### Steps for Testing


The focus should be code review, -> especially check if the changelog is correct and the code quality is good.
I did not add new functionality, but working functionality should not be broken with the enhancement of the data model.

1. Log in to Artemis
2. Navigate to an exercise or lecture of a course with enabled student posts
3. Create, Edit (own), delete (own) posts and answer posts as student
4. Create, Edit (arbitrary), delete (arbitrary), approve posts and answer posts as instructor

### Test Coverage




`PostResource.java`: 99%
`PostContextConstraintValidator.java`: 100%
`AnswerPostResource.java`: 94%
`ReactionResource.java`: 100%
`ReactionConstraintValidator.java`: 100%
